packages/vm/src/context.rs

```diff
@@ -62,229 +58,243 @@ impl GasState {
     }
 }
 
-struct ContextData<S: Storage, Q: Querier> {
-    gas_state: GasState,
-    storage: Option<S>,
-    storage_readonly: bool,
-    querier: Option<Q>,
-    /// A non-owning link to the wasmer instance
-    wasmer_instance: Option<NonNull<WasmerInstance>>,
+/// A ContextEnv is an env that provides access to the
+/// ContextData. The env is clonable but a clone accesses
+/// the same underlying data.
+pub struct Env<S: Storage, Q: Querier> {
+    data: Arc<RwLock<ContextData<S, Q>>>,
 }
 
-pub fn setup_context<S: Storage, Q: Querier>(gas_limit: u64) -> (*mut c_void, fn(*mut c_void)) {
-    (
-        create_unmanaged_context_data::<S, Q>(gas_limit),
-        destroy_unmanaged_context_data::<S, Q>,
-    )
+impl<S: Storage, Q: Querier> Clone for Env<S, Q> {
+    fn clone(&self) -> Self {
+        Env {
+            data: self.data.clone(),
+        }
+    }
 }
 
-fn create_unmanaged_context_data<S: Storage, Q: Querier>(gas_limit: u64) -> *mut c_void {
-    let data = ContextData::<S, Q> {
-        gas_state: GasState::with_limit(gas_limit),
-        storage: None,
-        storage_readonly: true,
-        querier: None,
-        wasmer_instance: None,
-    };
-    let heap_data = Box::new(data); // move from stack to heap
-    Box::into_raw(heap_data) as *mut c_void // give up ownership
+impl<S: Storage, Q: Querier> WasmerEnv for Env<S, Q> {
+    fn init_with_instance(&mut self, _instance: &WasmerInstance) -> Result<(), HostEnvInitError> {
+        Ok(())
+    }
 }
 
-fn destroy_unmanaged_context_data<S: Storage, Q: Querier>(ptr: *mut c_void) {
-    if !ptr.is_null() {
-        // obtain ownership and drop instance of ContextData when box gets out of scope
-        let _dying = unsafe { Box::from_raw(ptr as *mut ContextData<S, Q>) };
+impl<S: Storage, Q: Querier> Env<S, Q> {
+    pub fn new(gas_limit: u64) -> Self {
+        Env {
+            data: Arc::new(RwLock::new(ContextData::new(gas_limit))),
+        }
     }
-}
 
-/// Get a mutable reference to the context's data. Ownership remains in the Context.
-fn get_context_data_mut<'a, 'b, S: Storage, Q: Querier>(
-    ctx: &'a mut Ctx,
-) -> &'b mut ContextData<S, Q> {
-    unsafe {
-        let ptr = ctx.data as *mut ContextData<S, Q>;
-        ptr.as_mut()
-            .expect("The pointer ctx.data was null in get_context_data_mut; this is a bug.")
+    pub fn with_context_data_mut<Callback, CallbackReturn>(
```

Comment:
I see you got for longer, clearer generic names, rather than `T, U` common to Rust.

It does read nice in this case (which is also one of those confusing detail codes), but in general I prefer the short letters

Reply:
Yeah, agreed. This looked like types, not generics. Made it more compact in 81381c1c9864afc2e95328d9eee0b23acd5e0bf8